comm/src/main/scala/coop/rchain/p2p/p2p.scala

```diff
@@ -47,85 +47,133 @@ case object NetworkAddress {
 
 object Network extends ProtocolDispatcher[java.net.SocketAddress] {
 
-  type KeysStore[F[_]] = Kvs[F, PeerNode, Array[Byte]]
+  type KeysStore[F[_]]    = Kvs[F, PeerNode, Array[Byte]]
+  type ErrorHandler[F[_]] = ApplicativeError_[F, CommError]
 
   import NetworkProtocol._
   import Encryption._
 
+  val defaultTimeout: Duration = Duration(100, MILLISECONDS)
+
   def unsafeRoundTrip[F[_]: Capture: Communication]
     : (ProtocolMessage, ProtocolNode) => CommErr[ProtocolMessage] =
     (pm: ProtocolMessage, pn: ProtocolNode) => {
-      val result = Communication[F].roundTrip(pm, pn)
+      val result = Communication[F].roundTrip(pm, pn, defaultTimeout)
       Capture[F].unsafeUncapture(result)
     }
 
-  def connect[F[_]: Capture: Monad: Log: Time: Metrics: Communication: Encryption](peer: PeerNode)(
-      implicit
-      keysStore: KeysStore[F],
-      err: ApplicativeError_[F, CommError]): F[Unit] = {
+  def findAndConnect[
+      F[_]: Capture: Monad: Log: Time: Metrics: Communication: Encryption: KeysStore: ErrorHandler]
+    : Int => F[Int] =
+    (lastCount: Int) =>
+      (for {
+        _     <- IOUtil.sleep[F](5000L)
+        peers <- Communication[F].findMorePeers(10)
+        _ <- peers.toList
+              .traverse(p => errorHandler[F].attempt(connect[F](p, defaultTimeout)))
+              .map { attempts =>
+                attempts.filter {
+                  case Left(_) => false
+                  case _       => true
+                }
+              }
+        thisCount <- Communication[F].countPeers
+        _ <- if (thisCount != lastCount) Log[F].info(s"Peers: $thisCount.")
+            else ().pure[F]
+      } yield thisCount)
+
+  def connectToBootstrap[
+      F[_]: Capture: Monad: Log: Time: Metrics: Communication: Encryption: KeysStore: ErrorHandler](
+      bootstrapAddrStr: String,
+      maxNumOfAttempts: Int = 5): F[Unit] = {
+
+    def connectAttempt(attempt: Int, timeout: Duration, bootstrapAddr: PeerNode): F[Unit] =
+      if (attempt > maxNumOfAttempts) for {
+        _ <- Log[F].error("Failed to connect to bootstrap node, exiting...")
+        _ <- errorHandler[F].raiseError[Unit](couldNotConnectToBootstrap)
+      } yield ()
+      else
+        for {
+          res <- connect[F](bootstrapAddr, timeout).attempt
+          _ <- res match {
+                case Left(err) =>
```

Comment:
Is it worth attempting to reconnect on any kind of error? I know the number of attempts won't be big anyway so it doesn't hurt that match.

Reply:
It is because network is UDP based, we could might as well just timeouted. Read JIRA ticket for details.